org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/prolog/PrologModel.java

```diff
@@ -0,0 +1,201 @@
+		item.setInsertTextFormat(insertText);
+		int closingBracketOffset;
+		if (inPIState) {
+			closingBracketOffset = XMLCompletions.getOffsetFollowedBy(document.getText(), tokenEndOffset, ScannerState.WithinPI,
+					TokenType.PIEnd);
+		} else {// prolog state
+			closingBracketOffset = XMLCompletions.getOffsetFollowedBy(document.getText(), tokenEndOffset, ScannerState.WithinTag,
+					TokenType.PrologEnd);
+		}
+
+		if (closingBracketOffset != -1) {
+			// Include '?>'
+			closingBracketOffset += 2;
+		} else {
+			closingBracketOffset = XMLCompletions.getOffsetFollowedBy(document.getText(), tokenEndOffset, ScannerState.WithinTag,
+					TokenType.StartTagClose);
+			if (closingBracketOffset == -1) {
+				closingBracketOffset = tokenEndOffset;
+			} else {
+				closingBracketOffset++;
+			}
+		}
+		int startOffset = tokenEndOffset - tag.length();
+		try {
+			Range editRange = XMLCompletions.getReplaceRange(startOffset, closingBracketOffset, request);
+			String q = settings.formattingSettings.getQuotationAsString();
+			String cursor = isSnippetsSupported ? "$0" : "";
+			String text = MessageFormat.format("xml version={0}{1}{0} encoding={0}{2}{0}?>" + cursor , q, VERSION_1, UTF_8);
+			item.setTextEdit(new TextEdit(editRange, text));
+		} catch (BadLocationException e) {
+			LOGGER.log(Level.SEVERE, "While performing getReplaceRange for prolog completion.", e);
+		}
+		response.addCompletionItem(item);
+	}
+	
+
+	private static void createCompletionItem(String attrName, boolean canSupportSnippet, boolean generateValue,
+			Range editRange, String defaultValue, Collection<String> enumerationValues, String documentation,
+			ICompletionResponse response, SharedSettings settings){
+		CompletionItem item = new AttributeCompletionItem(attrName, canSupportSnippet, editRange, generateValue,
+				defaultValue, enumerationValues, settings);
+		MarkupContent markup = new MarkupContent();
+		markup.setKind(MarkupKind.MARKDOWN);
+		
+		markup.setValue(documentation);
+		item.setDocumentation(markup);
+		response.addCompletionItem(item);
+	}
+
+	public static void computeAttributeNameCompletionResponses(ICompletionRequest request, 
+	ICompletionResponse response, Range editRange, DOMDocument document, SharedSettings settings)
+			throws BadLocationException {
+
+		if (document.hasProlog() == false) {
+			return;
+		}
+		int offset = document.offsetAt(editRange.getStart());
+		DOMNode prolog = document.findNodeAt(offset);
+		if(!prolog.isProlog()) {
+			return;
+		}
+		boolean isSnippetsSupported = request.getCompletionSettings().isCompletionSnippetsSupported();
+		
+		if(!prolog.hasAttribute(VERSION_NAME)) {
+			createCompletionItem(VERSION_NAME, isSnippetsSupported, true, editRange, VERSION_1, VERSION_VALUES, null, response, settings);
+		}
+
+		if(!prolog.hasAttribute(ENCODING_NAME)) {
+			createCompletionItem(ENCODING_NAME, isSnippetsSupported, true, editRange, UTF_8, ENCODING_VALUES, null, response, settings);
+		}
```

Comment:
so apparently we should also propose a "standalone" attribute (yes/no values): https://www.ibm.com/support/knowledgecenter/en/SSMKHH_10.0.0/com.ibm.etools.mft.doc/ac67224_.htm

That would not be part of the the full prolog snippet though.